File: clients/php/src/Client/HttpEtwinClient.php
<?php declare(strict_types=1);

namespace Eternaltwin\Client;

use Eternaltwin\Auth\AccessTokenAuthContext;
use Eternaltwin\Auth\AuthContext;
use Eternaltwin\Auth\GuestAuthContext;
use Eternaltwin\Auth\UserAuthContext;
use Eternaltwin\User\User;
use Eternaltwin\User\UserId;
use GuzzleHttp\Client;
use GuzzleHttp\Psr7\Uri;
use GuzzleHttp\Psr7\UriResolver;

final class HttpEtwinClient implements EtwinClient {
  const DEFAULT_TIMEOUT = 30.0;

  private Uri $baseUri;
  private Client $client;


  final public function __construct(string $baseUri) {
    $parsedBaseUri = new Uri($baseUri);
    $inputPath = $parsedBaseUri->getPath();
    if (mb_substr($inputPath, -1, 1) != "/") {
      $parsedBaseUri = $parsedBaseUri->withPath($inputPath . "/");
    }
    $this->baseUri = $parsedBaseUri;
    $this->client = new Client(["base_uri" => $parsedBaseUri, "timeout" => self::DEFAULT_TIMEOUT]);
  }

  /**
   * @param Auth $auth
   * @return AccessTokenAuthContext | GuestAuthContext | UserAuthContext
   * @throws \JsonException
   */
  function getSelf(Auth $auth) {
    $res = $this->client->get(
      $this->resolve(["auth", "self"]),
      [
        "headers" => [
          "Authorization" => $auth->getAuthorizationHeader(),
        ],
      ],
    );
    $resBody = $res->getBody()->getContents();
    return AuthContext::fromJson($resBody);
  }

  function getUser(Auth $auth, UserId $userId): User {
    $res = $this->client->get(
      $this->resolve(["users", $userId->toString()]),
      [
        "headers" => [
          "Authorization" => $auth->getAuthorizationHeader(),
        ],
      ],
    );
    $resBody = $res->getBody()->getContents();
    return User::fromJson($resBody);
  }

  /**
   * @param string[] $segments
   * @return Uri
   */
  public function resolve(array $segments): Uri {
    return UriResolver::resolve($this->baseUri, new Uri("api/v1/" . implode("/", $segments)));
  }
}

File: clients/ruby/lib/etwin/client/http_etwin_client.rb
# typed: strict
# frozen_string_literal: true

module Etwin
  module Client
    # HTTP Etwin client interface
    class HttpEtwinClient
      extend T::Helpers
      extend T::Sig

      include EtwinClient

      final!

      sig(:final) { params(base_uri: URI::HTTP).void }
      def initialize(base_uri)
        # @type [URI::HTTP]
        @base_uri = T.let(base_uri.freeze, URI::HTTP)
        # @type [Faraday::Connection]
        @client = T.let(Faraday::Connection.new.freeze, Faraday::Connection)
        freeze
      end

      sig(:final) { override.params(auth: Auth, user_id: Etwin::User::UserId).returns(Etwin::User::User) }
      def get_user(auth, user_id)
        uri = resolve(['users', user_id.to_s])
        # @type [Faraday::Request] req
        # @type [Faraday::Response] res
        res = @client.get uri do |req|
          req.headers['Authorization'] = auth.authorization_header
        end
        Etwin::User::User.from_json res.body
      end

      private

      sig(:final) { params(segments: T::Array[String]).returns(URI::HTTP) }
      def resolve(segments)
        @base_uri.merge("api/v1/#{ segments * '/'}")
      end
    end
  end
end
